File: src/main/services/comicinfo.ts
/**
 * ComicInfo.xml — the parse side.
 *
 * Writing is no longer done here. The bytes come from
 * `resources/metadata-templates/comicinfo.template`, mapped from a
 * `FileMetadata` by `services/metadata/mappers.ts`, so that what this app
 * writes can be changed by editing a text file. Reading stays in code: it has
 * to cope with every shape this library has accumulated over the years,
 * including files written by other tools.
 *
 * Pure and synchronous — no file or ZIP access, so it is trivially testable.
 */

import { decodeXmlEntities } from './xml-utils'

// ─── Types ───────────────────────────────────────────────────────────────────

/**
 * What `parseComicInfoXml` can recover from a ComicInfo.xml.
 *
 * A read-side shape only. The write side starts from `FileMetadata`, which
 * carries considerably more — see `services/metadata/file-metadata.ts`.
 */
export interface ComicInfoMetadata {
  title: string
  series: string
  /** Position within the series, read from Number or from a legacy Volume. */
  volume?: number | null
  summary?: string | null
  writers: string[]
  publisher?: string | null
  /** nhentai category tags. */
  genres: string[]
  /** nhentai tag-type tags. */
  tags: string[]
  characters: string[]
  webUrl?: string | null
  notes?: string | null
  pageCount: number
  languageIso?: string | null
  releaseDate?: Date | null
  ageRating: string
  manga: 'Yes' | 'YesAndRightToLeft' | 'No'
  /** Optional parody, which Kavita treats as a collection. */
  seriesGroup?: string | null
}

// ─── Parse ────────────────────────────────────────────────────────────────────

/**
 * Parse a ComicInfo.xml string into a Partial<ComicInfoMetadata>.
 *
 * Uses per-field regex extraction rather than a full XML parser (the schema
 * is flat — no nesting except optional <Pages>, which we do not read).
 * Entities are decoded so that e.g. `&` round-trips to `&` (§4.7).
 *
 * Returns a partial object — only fields found in the XML are set.
 */
export function parseComicInfoXml(xml: string): Partial<ComicInfoMetadata> {
  const result: Partial<ComicInfoMetadata> = {}

  const text = (raw: string): string => decodeXmlEntities(raw.trim())

  const extract = (tag: string): string | null => {
    const re = new RegExp(`<${tag}[^>]*>([^<]+)</${tag}>`, 'i')
    const m = xml.match(re)
    return m ? text(m[1]) : null
  }

  const title = extract('Title')
  if (title) result.title = title

  const series = extract('Series')
  if (series) result.series = series

  /*
   * Position within the series, from Number first and Volume second.
   *
   * The writer emits Number. Volume is still read because every file written
   * before that change used it, and the scanner takes `seriesIndex` from here
   * — reading only Number would drop the index from the whole existing library
   * on the next rescan, and reading only Volume would drop it from everything
   * written from now on.
   */
  const indexStr = extract('Number') || extract('Volume')
  if (indexStr) {
    const parsed = parseFloat(indexStr)
    if (!isNaN(parsed)) result.volume = parsed
  }

  const summary = extract('Summary')
  if (summary) result.summary = summary

  const writers = extract('Writer')
  if (writers) {
    result.writers = writers.split(',').map((s) => s.trim()).filter(Boolean)
  }

  const publisher = extract('Publisher')
  if (publisher) result.publisher = publisher

  const genres = extract('Genre')
  if (genres) {
    result.genres = genres.split(',').map((s) => s.trim()).filter(Boolean)
  }

  const tags = extract('Tags')
  if (tags) {
    result.tags = tags.split(',').map((s) => s.trim()).filter(Boolean)
  }

  const characters = extract('Characters')
  if (characters) {
    result.characters = characters.split(',').map((s) => s.trim()).filter(Boolean)
  }

  const webUrl = extract('Web')
  if (webUrl) result.webUrl = webUrl

  const notes = extract('Notes')
  if (notes) result.notes = notes

  const pageCountStr = extract('PageCount')
  if (pageCountStr) {
    const parsed = parseInt(pageCountStr, 10)
    if (!isNaN(parsed)) result.pageCount = parsed
  }

  const languageIso = extract('LanguageISO')
  if (languageIso) result.languageIso = languageIso

  // Year/Month/Day — only if Year is present
  const yearStr = extract('Year')
  if (yearStr) {
    const y = parseInt(yearStr, 10)
    const mStr = extract('Month')
    const dStr = extract('Day')
    const m = mStr ? parseInt(mStr, 10) : 1
    const d = dStr ? parseInt(dStr, 10) : 1
    if (!isNaN(y)) {
      result.releaseDate = new Date(y, Math.max(0, (m || 1) - 1), d || 1)
    }
  }

  const ageRating = extract('AgeRating')
  if (ageRating) result.ageRating = ageRating as ComicInfoMetadata['ageRating']

  const manga = extract('Manga')
  if (manga && (manga === 'Yes' || manga === 'YesAndRightToLeft' || manga === 'No')) {
    result.manga = manga
  }

  const seriesGroup = extract('SeriesGroup')
  if (seriesGroup) result.seriesGroup = seriesGroup

  return result
}

File: kopibon-core/src/scanner/extract/comicinfo.rs
//! parseComicInfoXml port (src/main/services/comicinfo.ts:59+). Per-field
//! regexes rather than a full parser (flat schema), entity-decoded. Pure and
//! synchronous — the read side of the writer already in `metadata::writers`.

use serde_json::json;

use crate::metadata::xml_utils::decode_xml_entities as decode_shared;

/// parseComicInfoXml uses the SHARED xml-utils decode (comicinfo.ts:11) —
/// the `&#39;` variant, unlike the worker's own `&apos;` copy. A decode
/// failure (RangeError case, unreachable for real payloads) skips the field;
/// 1.x would throw into extractCbzMetadata's catch and drop the whole file's
/// metadata — documented divergence, same reason as xml_utils.rs:36-40.
fn decode_xml_entities(s: &str) -> Option<String> {
    decode_shared(s).ok()
}

/// What `parseComicInfoXml` can recover (comicinfo.ts:21-45) — the fields the
/// scanner consumes; the rest of the schema is write-side only.
#[derive(Debug, Default, Clone)]
pub struct ComicInfoMetadata {
    pub title: Option<String>,
    pub series: Option<String>,
    /// Position within the series, Number first then legacy Volume.
    pub volume: Option<f64>,
    pub summary: Option<String>,
    pub writers: Vec<String>,
    pub publisher: Option<String>,
    pub genres: Vec<String>,
    pub tags: Vec<String>,
    pub characters: Vec<String>,
    pub web_url: Option<String>,
    pub notes: Option<String>,
    pub page_count: Option<i64>,
    pub language_iso: Option<String>,
}

fn extract(tag: &str, xml: &str) -> Option<String> {
    let re = regex::Regex::new(&format!(
        r"(?i)<{tag}[^>]*>([^<]+)</{tag}>"
    ))
    .ok()?;
    let m = re.captures(xml)?;
    let raw = m.get(1)?.as_str();
    // text() = decodeXmlEntities(raw.trim())
    decode_xml_entities(raw.trim())
}

fn split_list(value: Option<String>) -> Vec<String> {
    value
        .map(|v| {
            v.split(',')
                .map(|s| s.trim().to_string())
                .filter(|s| !s.is_empty())
                .collect()
        })
        .unwrap_or_default()
}

pub fn parse_comic_info_xml(xml: &str) -> ComicInfoMetadata {
    let mut result = ComicInfoMetadata::default();

    if let Some(title) = extract("Title", xml) {
        if !title.is_empty() {
            result.title = Some(title);
        }
    }
    if let Some(series) = extract("Series", xml) {
        if !series.is_empty() {
            result.series = Some(series);
        }
    }
    // Number first, Volume second (comicinfo.ts:74-85).
    let index_str = extract("Number", xml).or_else(|| extract("Volume", xml));
    if let Some(index_str) = index_str {
        if let Some(parsed) = crate::scanner::extract::pdf::parse_js_float(&index_str) {
            if !parsed.is_nan() {
                result.volume = Some(parsed);
            }
        }
    }
    if let Some(summary) = extract("Summary", xml) {
        if !summary.is_empty() {
            result.summary = Some(summary);
        }
    }
    result.writers = split_list(extract("Writer", xml));
    if let Some(publisher) = extract("Publisher", xml) {
        if !publisher.is_empty() {
            result.publisher = Some(publisher);
        }
    }
    result.genres = split_list(extract("Genre", xml));
    result.tags = split_list(extract("Tags", xml));
    result.characters = split_list(extract("Characters", xml));
    if let Some(web) = extract("Web", xml) {
        if !web.is_empty() {
            result.web_url = Some(web);
        }
    }
    if let Some(notes) = extract("Notes", xml) {
        if !notes.is_empty() {
            result.notes = Some(notes);
        }
    }
    if let Some(pc) = extract("PageCount", xml) {
        if let Ok(parsed) = pc.parse::<i64>() {
            result.page_count = Some(parsed);
        }
    }
    if let Some(iso) = extract("LanguageISO", xml) {
        if !iso.is_empty() {
            result.language_iso = Some(iso);
        }
    }
    result
}

/// The gallery stub's `raw_tags_json` payload (upsertGalleryStub, :739).
/// Built as a raw string: `JSON.stringify({id, type, name})` keeps insertion
/// order, and serde_json's default BTreeMap would sort it (id, name, type) —
/// a byte-level divergence in a stored column.
pub fn stub_tags_json(tags: &[String]) -> String {
    let items: Vec<String> = tags
        .iter()
        .map(|t| {
            format!(
                "{{\"id\":0,\"type\":\"tag\",\"name\":{}}}",
                serde_json::to_string(t).unwrap_or_else(|_| "\"\"".to_string())
            )
        })
        .collect();
    format!("[{}]", items.join(","))
}

/// The gallery stub's `raw_json` payload (:740).
pub fn stub_raw_json(gallery_id: i64, title: &str) -> String {
    serde_json::to_string(&json!({ "id": gallery_id, "title": { "pretty": title } }))
        .unwrap_or_else(|_| "{}".to_string())
}
